feature/experimental/p4rt/ate_tests/performance_test/performance_test.go

```diff
@@ -0,0 +1,946 @@
+	GetPacketTemplate() *PacketIOPacket
+	GetEgressPort() string
+	GetIngressPort() string
+}
+
+type LLDPPacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+type GDPPacketIO struct {
+	PacketIOPacket
+	IngressPort string
+}
+
+type TraceroutePacketIO struct {
+	PacketIOPacket
+	IngressPort string
+	EgressPort  string
+}
+
+// programTableEntry programs or deletes p4rt table entry based on delete flag.
+func programTableEntry(client *p4rt_client.P4RTClient, delete bool) error {
+	return client.Write(&p4_v1.WriteRequest{
+		DeviceId:   deviceID,
+		ElectionId: &p4_v1.Uint128{High: uint64(0), Low: electionID},
+		Updates: p4rtutils.ACLWbbIngressTableEntryGet(
+			getTableEntry(delete),
+		),
+		Atomicity: p4_v1.WriteRequest_CONTINUE_ON_ERROR,
+	})
+}
+
+// testTraffic sends ATE traffic, stop and collect total packet tx from ATE source port.
+func testTraffic(t *testing.T, ate *ondatra.ATEDevice, flows []*ondatra.Flow, srcEndPoint *ondatra.Interface, duration int) int {
+	t.Helper()
+	for _, flow := range flows {
+		flow.WithSrcEndpoints(srcEndPoint).WithDstEndpoints(srcEndPoint)
+	}
+	ate.Traffic().Start(t, flows...)
+	time.Sleep(time.Duration(duration) * time.Second)
+	ate.Traffic().Stop(t)
+
+	outPkts := gnmi.GetAll(t, ate, gnmi.OC().FlowAny().Counters().OutPkts().State())
+	total := 0
+	for _, count := range outPkts {
+		total += int(count)
+	}
+	return total
+}
+
+// sendPackets GDP/LLDP/Traceroute sends out packets via PacketOut message in StreamChannel.
+func sendPackets(t *testing.T, client *p4rt_client.P4RTClient, packet *p4_v1.PacketOut, packetCount int, timer float64) {
+
+	for i := 0; i < packetCount; i++ {
+		if err := client.StreamChannelSendMsg(
+			&streamName, &p4_v1.StreamMessageRequest{
+				Update: &p4_v1.StreamMessageRequest_Packet{
+					Packet: packet,
+				},
+			}); err != nil {
+			t.Errorf("There is error seen in Packet Out. %v, %s", err, err)
+		}
+		time.Sleep(time.Duration(timer) * time.Millisecond)
+	}
+}
+
+// decodeIPPacketTTL decodes L2 header in the packet and returns TTL, packetData[14:0] to remove first 14 bytes of Ethernet header.
+func decodeIPPacketTTL(t *testing.T, packetData []byte) uint8 {
+	t.Helper()
+	packet := gopacket.NewPacket(packetData, layers.LayerTypeEthernet, gopacket.Default)
+	etherHeader := packet.Layer(layers.LayerTypeEthernet)
+	if etherHeader != nil {
+		header := etherHeader.(*layers.Ethernet)
+		if header.EthernetType == layers.EthernetType(0x0800) {
+			packet1 := gopacket.NewPacket(packetData[14:], layers.LayerTypeIPv4, gopacket.Default)
+			IPv4 := packet1.Layer(layers.LayerTypeIPv4)
+			if IPv4 != nil {
+				ipv4 := IPv4.(*layers.IPv4)
+				IPv4 := ipv4.TTL
+				return IPv4
+			}
+		}
+	}
+	return 7 // Return 7 if not a valid packet and no ethernet header present.
+}
+
+// decodeL2Packet decodes L2 header in the packet and returns destination MAC and ethernet type.
+func decodeL2Packet(t *testing.T, packetData []byte) (string, layers.EthernetType) {
+	t.Helper()
```

Comment:
this is the only use of t in the function, so this call can be dropped and the t parameter removed